{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b92e82a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "from Bio import Entrez, SeqIO\n",
    "import tqdm\n",
    "from collections import Counter\n",
    "import numpy as np\n",
    "from scipy.stats import multinomial\n",
    "import numpy as np\n",
    "import ast\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e7abfd34",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading data: 16858it [01:34, 178.11it/s]\n"
     ]
    }
   ],
   "source": [
    "# Load the data\n",
    "iterator = pd.read_csv(\"/Users/reem/Mov/final_results.tsv\", sep=\"\\t\", usecols = ['seqName','privateNucMutations.unlabeledSubstitutions'], dtype={\"seqName\":str, 'privateNucMutations.unlabeledSubstitutions':str } ,chunksize=1000)\n",
    "df = pd.concat([chunk for chunk in tqdm.tqdm(iterator, desc='Loading data')])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bede5b0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "spectra = [\n",
    "    {\"name\":\"BA.1\",\n",
    "     \"url\": \"https://raw.githubusercontent.com/theosanderson/molnupiravir/main/mutational_spectra/BA.1_SBS_spectrum_Ruis.csv\"\n",
    "    },\n",
    "    {\"name\":\"High G-to-A\",\n",
    "        \"url\": \"https://raw.githubusercontent.com/theosanderson/molnupiravir/main/mutational_spectra/long_phylogenetic_branches/long_branch_spectrum_rescaled.csv\"    \n",
    "    },\n",
    "]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "79e1ed8a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0                         [A2396G, T13832C, C25665T, C27673T]\n",
       "1           [C4901T, C6525T, A12162G, C12781T, G16558T, C2...\n",
       "2           [C4084T, T7092C, A13956G, C14585T, A20425G, C2...\n",
       "3                                                   [A29188G]\n",
       "4                                          [C21595T, G21777T]\n",
       "                                  ...                        \n",
       "16857147                          [C15540T, C25710T, G27870T]\n",
       "16857148    [C44T, -11296G, C12085T, C19164A, C21459A, G27...\n",
       "16857149                  [G4583A, C10376T, C26895T, C29272T]\n",
       "16857150                   [C2094T, C7528T, G21255T, C27657T]\n",
       "16857151                            [C609T, T21610C, C25611T]\n",
       "Name: privateNucMutations.unlabeledSubstitutions, Length: 16857152, dtype: object"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['privateNucMutations.unlabeledSubstitutions'].str.split(\",\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0013aabd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_reference_genome(accession='NC_045512.2'):\n",
    "    Entrez.email = \"theo@theo.io\"  \n",
    "    handle = Entrez.efetch(db=\"nucleotide\", id=accession, rettype=\"fasta\", retmode=\"text\")\n",
    "    record = SeqIO.read(handle, \"fasta\")\n",
    "    handle.close()\n",
    "    return str(record.seq)\n",
    "reference_genome = fetch_reference_genome(\"NC_045512.2\")  # SARS-CoV-2 reference genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "25b0dbde",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0                    A>G,T>C,C>T,C>T\n",
       "1    C>T,C>T,A>G,C>T,G>T,C>T,G>T,C>T\n",
       "2    C>T,T>C,A>G,C>T,A>G,C>T,C>T,C>T\n",
       "3                                A>G\n",
       "4                            C>T,G>T\n",
       "Name: subs, dtype: object"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_mut_type(mut_string):\n",
    "    if isinstance(mut_string, str) and len(mut_string) >= 2:\n",
    "        return mut_string[0] + '>' + mut_string[-1]\n",
    "    else:\n",
    "        return ''\n",
    "df[\"subs\"]=df[\"privateNucMutations.unlabeledSubstitutions\"].apply(lambda x: ','.join([get_mut_type(item) for item in x.split(',')])if isinstance(x, str) else '')\n",
    "df[\"subs\"].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9cbd1343",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0                              {'A>G': 1, 'T>C': 1, 'C>T': 2}\n",
       "1                              {'C>T': 5, 'A>G': 1, 'G>T': 2}\n",
       "2                              {'C>T': 5, 'T>C': 1, 'A>G': 2}\n",
       "3                                                  {'A>G': 1}\n",
       "4                                        {'C>T': 1, 'G>T': 1}\n",
       "                                  ...                        \n",
       "16857147                                 {'C>T': 2, 'G>T': 1}\n",
       "16857148    {'C>T': 2, '->G': 1, 'C>A': 2, 'G>T': 1, 'A>G'...\n",
       "16857149                                 {'G>A': 1, 'C>T': 3}\n",
       "16857150                                 {'C>T': 3, 'G>T': 1}\n",
       "16857151                                 {'C>T': 2, 'T>C': 1}\n",
       "Name: Counts, Length: 16857152, dtype: object"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"Counts\"] =df[\"subs\"].apply(lambda x: dict(Counter(x.split(\",\"))))\n",
    "df[\"Counts\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9bdacd20",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0                               CAC,TTA,ACT,TCA\n",
       "1               CCT,ACA,CAG,ACA,TGC,ACA,TGG,CCA\n",
       "2               ACA,GTC,TAC,GCT,TAT,TCG,TCT,ACA\n",
       "3                                           CAC\n",
       "4                                       TCT,GGG\n",
       "                           ...                 \n",
       "16857147                            TCA,TCT,TGA\n",
       "16857148    TCG,TTA,CCT,ACA,TCA,TGG,AAT,AAA,AAA\n",
       "16857149                        CGA,ACC,CCA,ACA\n",
       "16857150                        TCG,TCG,CGT,TCA\n",
       "16857151                            CCA,TTA,TCT\n",
       "Name: context, Length: 16857152, dtype: object"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_context(genome_seq, mutation):\n",
    "    if isinstance(mutation, str) and len(mutation) >= 2:\n",
    "        pos = int(mutation[1:-1]) - 1  # -1 as Python uses 0-based indexing\n",
    "        context = genome_seq[pos-1:pos+2]  # get the base before and after\n",
    "        return context\n",
    "    else:\n",
    "        return ''\n",
    "df[\"context\"]=df[\"privateNucMutations.unlabeledSubstitutions\"].apply(lambda x: ','.join([get_context(reference_genome, item) for item in x.split(',')])if isinstance(x, str) else '')\n",
    "df[\"context\"]   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "40fef91b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqName</th>\n",
       "      <th>privateNucMutations.unlabeledSubstitutions</th>\n",
       "      <th>subs</th>\n",
       "      <th>Counts</th>\n",
       "      <th>context</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>hCoV-19/Czech Republic/NAT-22-02478/2022|2022-...</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>{'': 1}</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>hCoV-19/USA/NY-NYCPHL-016290/2022|2022-12-15|2...</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>{'': 1}</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>hCoV-19/USA/CT-CDC-QDX26076006/2021|2021-06-19...</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>{'': 1}</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>hCoV-19/Japan/PG-118664/2021|2021-08-24|2021-1...</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>{'': 1}</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>hCoV-19/Canada/BC-BCCDC-683663/2023|2023-12-06...</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>{'': 1}</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              seqName  \\\n",
       "17  hCoV-19/Czech Republic/NAT-22-02478/2022|2022-...   \n",
       "30  hCoV-19/USA/NY-NYCPHL-016290/2022|2022-12-15|2...   \n",
       "35  hCoV-19/USA/CT-CDC-QDX26076006/2021|2021-06-19...   \n",
       "38  hCoV-19/Japan/PG-118664/2021|2021-08-24|2021-1...   \n",
       "41  hCoV-19/Canada/BC-BCCDC-683663/2023|2023-12-06...   \n",
       "\n",
       "   privateNucMutations.unlabeledSubstitutions subs   Counts context  \n",
       "17                                        NaN       {'': 1}          \n",
       "30                                        NaN       {'': 1}          \n",
       "35                                        NaN       {'': 1}          \n",
       "38                                        NaN       {'': 1}          \n",
       "41                                        NaN       {'': 1}          "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['context'].apply(lambda x: isinstance(x, str) and len(x) < 2)].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "464cd772",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0                      C[A>G]C,T[T>C]A,A[C>T]T,T[C>T]A\n",
       "1    C[C>T]T,A[C>T]A,C[A>G]G,A[C>T]A,T[G>T]C,A[C>T]...\n",
       "2    A[C>T]A,G[T>C]C,T[A>G]C,G[C>T]T,T[A>G]T,T[C>T]...\n",
       "3                                              C[A>G]C\n",
       "4                                      T[C>T]T,G[G>T]G\n",
       "Name: spectrum, dtype: object"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def spectrum(subs, contexts):\n",
    "    if not isinstance(subs, str) or not isinstance(contexts, str):\n",
    "        return ''\n",
    "    if not subs.strip() or not contexts.strip():\n",
    "        return ''\n",
    "    subs= subs.split(',')\n",
    "    contexts = contexts.split(',')\n",
    "    spectra = []\n",
    "    for mutation, context in zip(subs, contexts): \n",
    "        if len(context)<2:\n",
    "            continue\n",
    "        spectra.append(f\"{context[0]}[{mutation}]{context[-1]}\")\n",
    "    return ','.join(spectra)\n",
    "   \n",
    "df[\"spectrum\"] = df.apply(lambda row: spectrum(row[\"subs\"], row[\"context\"]), axis=1)\n",
    "df[\"spectrum\"].head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a16a18b7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0    {}\n",
      "1    {}\n",
      "2    {}\n",
      "3    {}\n",
      "4    {}\n",
      "Name: G>A_counts, dtype: object\n",
      "0                  {'C[A>G]C': 1}\n",
      "1                  {'C[A>G]G': 1}\n",
      "2    {'T[A>G]C': 1, 'T[A>G]T': 1}\n",
      "3                  {'C[A>G]C': 1}\n",
      "4                              {}\n",
      "Name: A>G_counts, dtype: object\n",
      "0                         {'A[C>T]T': 1, 'T[C>T]A': 1}\n",
      "1           {'C[C>T]T': 1, 'A[C>T]A': 3, 'C[C>T]A': 1}\n",
      "2    {'A[C>T]A': 2, 'G[C>T]T': 1, 'T[C>T]G': 1, 'T[...\n",
      "3                                                   {}\n",
      "4                                       {'T[C>T]T': 1}\n",
      "Name: C>T_counts, dtype: object\n",
      "0    {'T[T>C]A': 1}\n",
      "1                {}\n",
      "2    {'G[T>C]C': 1}\n",
      "3                {}\n",
      "4                {}\n",
      "Name: T>C_counts, dtype: object\n"
     ]
    }
   ],
   "source": [
    "#Get Counts per substitution context\n",
    "def count_GtoA(spectrum):\n",
    "    counts = Counter()\n",
    "    if not isinstance(spectrum, str):\n",
    "        return {}\n",
    "    muts = spectrum.split(\",\")\n",
    "    for mut in muts:\n",
    "        if mut[2:5] == 'G>A':\n",
    "            counts[mut]+=1\n",
    "    return counts\n",
    "\n",
    "def count_AtoG(spectrum):\n",
    "    counts = Counter()\n",
    "    if not isinstance(spectrum, str):\n",
    "        return {}\n",
    "    muts = spectrum.split(\",\")\n",
    "    for mut in muts:\n",
    "        if mut[2:5] == 'A>G':\n",
    "            counts[mut]+=1\n",
    "    return counts\n",
    "\n",
    "def count_CtoT(spectrum):\n",
    "    counts = Counter()\n",
    "    if not isinstance(spectrum, str):\n",
    "        return {}\n",
    "    muts = spectrum.split(\",\")\n",
    "    for mut in muts:\n",
    "        if mut[2:5] == 'C>T':\n",
    "            counts[mut]+=1\n",
    "    return counts\n",
    "\n",
    "def count_TtoC(spectrum):\n",
    "    counts = Counter()\n",
    "    if not isinstance(spectrum, str):\n",
    "        return {}\n",
    "    muts = spectrum.split(\",\")\n",
    "    for mut in muts:\n",
    "        if mut[2:5] == 'T>C':\n",
    "            counts[mut]+=1\n",
    "    return counts\n",
    "\n",
    "df[\"G>A_counts\"] = df[\"spectrum\"].apply(count_GtoA)\n",
    "df[\"A>G_counts\"] = df[\"spectrum\"].apply(count_AtoG)\n",
    "df[\"C>T_counts\"] = df[\"spectrum\"].apply(count_CtoT)\n",
    "df[\"T>C_counts\"] = df[\"spectrum\"].apply(count_TtoC)\n",
    "print(df[\"G>A_counts\"].head())\n",
    "print(df[\"A>G_counts\"].head())\n",
    "print(df[\"C>T_counts\"].head())\n",
    "print(df[\"T>C_counts\"].head())\n",
    "\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "ff7d01dc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0    {}\n",
      "1    {}\n",
      "2    {}\n",
      "3    {}\n",
      "4    {}\n",
      "Name: G>Aproportions, dtype: object\n",
      "0                    {'C[A>G]C': 1.0}\n",
      "1                    {'C[A>G]G': 1.0}\n",
      "2    {'T[A>G]C': 0.5, 'T[A>G]T': 0.5}\n",
      "3                    {'C[A>G]C': 1.0}\n",
      "4                                  {}\n",
      "Name: A>Gproportions, dtype: object\n",
      "0                     {'A[C>T]T': 0.5, 'T[C>T]A': 0.5}\n",
      "1     {'C[C>T]T': 0.2, 'A[C>T]A': 0.6, 'C[C>T]A': 0.2}\n",
      "2    {'A[C>T]A': 0.4, 'G[C>T]T': 0.2, 'T[C>T]G': 0....\n",
      "3                                                   {}\n",
      "4                                     {'T[C>T]T': 1.0}\n",
      "Name: C>Tproportions, dtype: object\n",
      "0    {'T[T>C]A': 1.0}\n",
      "1                  {}\n",
      "2    {'G[T>C]C': 1.0}\n",
      "3                  {}\n",
      "4                  {}\n",
      "Name: T>Cproportions, dtype: object\n"
     ]
    }
   ],
   "source": [
    "# Get Proportions per substitution context\n",
    "def get_proportion(df):\n",
    "    dict = {}\n",
    "    total = sum(df.values())\n",
    "    for key, value in df.items():\n",
    "        dict[key] = value/total\n",
    "       \n",
    "    return dict\n",
    "\n",
    "df[\"G>Aproportions\"] = df.apply(lambda row: get_proportion(row[\"G>A_counts\"]), axis=1)\n",
    "df[\"A>Gproportions\"] = df.apply(lambda row: get_proportion(row[\"A>G_counts\"]), axis=1)\n",
    "df[\"C>Tproportions\"] = df.apply(lambda row: get_proportion(row[\"C>T_counts\"]), axis=1)\n",
    "df[\"T>Cproportions\"] = df.apply(lambda row: get_proportion(row[\"T>C_counts\"]), axis=1)\n",
    "\n",
    "print(df[\"G>Aproportions\"].head())\n",
    "print(df[\"A>Gproportions\"].head())\n",
    "print(df[\"C>Tproportions\"].head())\n",
    "print(df[\"T>Cproportions\"].head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "ac15c719",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0               {'A>G': 1, 'T>C': 1, 'C>T': 2}\n",
      "1               {'C>T': 5, 'A>G': 1, 'G>T': 2}\n",
      "2               {'C>T': 5, 'T>C': 1, 'A>G': 2}\n",
      "3                                   {'A>G': 1}\n",
      "4                         {'C>T': 1, 'G>T': 1}\n",
      "5               {'G>T': 6, 'C>T': 3, 'A>G': 1}\n",
      "6     {'T>G': 1, 'T>C': 1, 'C>T': 2, 'G>A': 1}\n",
      "7               {'C>T': 5, '->G': 1, 'T>C': 1}\n",
      "8                         {'T>C': 1, 'G>T': 1}\n",
      "9                         {'T>C': 1, 'C>T': 1}\n",
      "10              {'C>T': 2, 'A>T': 1, 'T>C': 2}\n",
      "11              {'G>T': 1, 'T>C': 1, 'C>T': 2}\n",
      "12                        {'G>T': 1, 'G>A': 1}\n",
      "13    {'G>T': 1, 'C>T': 4, 'T>C': 1, 'G>A': 1}\n",
      "14              {'A>G': 2, 'C>A': 1, 'C>T': 2}\n",
      "15                        {'C>T': 2, 'A>G': 1}\n",
      "16              {'C>T': 3, 'T>C': 2, 'G>T': 1}\n",
      "17                                     {'': 1}\n",
      "18    {'G>T': 2, 'T>C': 2, 'C>T': 1, 'G>A': 1}\n",
      "19              {'C>T': 4, 'G>T': 1, 'T>C': 2}\n",
      "Name: Counts, dtype: object\n",
      "<class 'pandas.core.series.Series'>\n"
     ]
    }
   ],
   "source": [
    "# Because df_Counts is not read as a dict in the next step\n",
    "\n",
    "df[\"Counts\"] = (\n",
    "    df[\"Counts\"]\n",
    "    .astype(str)\n",
    "    .str.replace(\"Counter(\", \"\")  # remove \"Counter(\"\n",
    "    .str.rstrip(\")\")                           # remove trailing \")\"\n",
    "    .apply(lambda x: ast.literal_eval(x) if x not in [\"\", \"nan\", \"Counter\"] else {})\n",
    ")\n",
    "print(df[\"Counts\"].head(20))\n",
    "print(type(df[\"Counts\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "be9748a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    0.133011\n",
       "1   -1.049009\n",
       "2    1.231103\n",
       "3   -0.210447\n",
       "4   -1.073550\n",
       "Name: LLR, dtype: float64"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_likelihood_ratio(counts,pM,pN):\n",
    "    counts=np.array(counts,dtype=float)\n",
    "    llM= float(multinomial.logpmf(counts, n=np.sum(counts), p=pM))\n",
    "    llN = float(multinomial.logpmf(counts, n=np.sum(counts), p=pN))\n",
    "    llr=llM-llN\n",
    "    return llr\n",
    "probs_df = pd.read_csv(\"/Users/reem/Downloads/estimated_mutation_distribution.tsv\", delimiter=\"\\t\")\n",
    "#probs_df\n",
    "\n",
    "pM=probs_df[\"Molnupiravir\"].values.tolist()\n",
    "pN=probs_df[\"Normal\"].values.tolist()\n",
    "mutation_types = probs_df[\"MutationType\"].str.replace(\"â†’\", \">\").tolist()\n",
    "\n",
    "\n",
    "llr_list=[]\n",
    "for i,row in df.iterrows():\n",
    "    counts_dict = row[\"Counts\"]\n",
    "    counts = [counts_dict.get(mut,0) for mut in mutation_types]\n",
    "    llr = get_likelihood_ratio(counts,pM,pN)\n",
    "    llr_list.append(llr)\n",
    "df[\"LLR\"] = llr_list\n",
    "df[\"LLR\"].head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "a1d6f497",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-6208.81062121423\n",
      "31.137873738090548\n"
     ]
    }
   ],
   "source": [
    "print(df[\"LLR\"].min())\n",
    "print(df[\"LLR\"].max())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "100b28d6",
   "metadata": {},
   "outputs": [
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mThe Kernel crashed while executing code in the current cell or a previous cell. \n",
      "\u001b[1;31mPlease review the code in the cell(s) to identify a possible cause of the failure. \n",
      "\u001b[1;31mClick <a href='https://aka.ms/vscodeJupyterKernelCrash'>here</a> for more info. \n",
      "\u001b[1;31mView Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details."
     ]
    }
   ],
   "source": [
    "df.to_csv(\"/Users/reem/Mov/final_results_LLRs.tsv\",sep=\"\\t\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
