{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b92e82a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "from Bio import Entrez, SeqIO\n",
    "import tqdm\n",
    "from collections import Counter\n",
    "import numpy as np\n",
    "from scipy.stats import multinomial\n",
    "import numpy as np\n",
    "import ast\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e7abfd34",
   "metadata": {},
   "outputs": [],
   "source": [
    "file = \"/Users/reem/Mov/llr_top_1000.tsv\"\n",
    "df =pd.read_csv(file, delimiter=\"\\t\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bede5b0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "spectra = [\n",
    "    {\"name\":\"BA.1\",\n",
    "     \"url\": \"https://raw.githubusercontent.com/theosanderson/molnupiravir/main/mutational_spectra/BA.1_SBS_spectrum_Ruis.csv\"\n",
    "    },\n",
    "    {\"name\":\"High G-to-A\",\n",
    "        \"url\": \"https://raw.githubusercontent.com/theosanderson/molnupiravir/main/mutational_spectra/long_phylogenetic_branches/long_branch_spectrum_rescaled.csv\"    \n",
    "    },\n",
    "]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "79e1ed8a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      [G204A, C2445T, C4331T, C5621T, C5622T, C6633T...\n",
       "1                                                    NaN\n",
       "2                                                    NaN\n",
       "3                                              [G19999T]\n",
       "4                                        [G410T, A7881G]\n",
       "                             ...                        \n",
       "994                                             [A4759G]\n",
       "995                          [C11962T, C13694T, T18132C]\n",
       "996    [G3692T, G6476T, T13188C, T13626C, G16528A, G1...\n",
       "997    [G922A, C6428T, C8802T, G11108T, A18146G, G202...\n",
       "998                   [C1441T, A1732G, G27517A, G29734T]\n",
       "Name: privateNucMutations.unlabeledSubstitutions, Length: 999, dtype: object"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['privateNucMutations.unlabeledSubstitutions'].str.split(\",\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "07e1edda",
   "metadata": {},
   "outputs": [],
   "source": [
    "def GtoA_percent(muts):\n",
    "    mut_list=muts.split(\",\")\n",
    "    total=len(mut_list)\n",
    "    total_muts=0\n",
    "    G_to_A=0\n",
    "    for m in mut_list:\n",
    "        total_muts+=1\n",
    "        if m.startswith(\"G\") and m.endswith(\"A\"):\n",
    "            G_to_A+=1\n",
    "    if total > 0:\n",
    "        return((G_to_A/total_muts) * 100)\n",
    "#df = pd.DataFrame(df)\n",
    "#df[\"G_to_A_%\"] = df[\"privateNucMutations.unlabeledSubstitutions\"].apply(GtoA_percent)\n",
    "#print(df[[\"seqName\", \"privateNucMutations.unlabeledSubstitutions\", \"G_to_A_%\"]])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0013aabd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_reference_genome(accession='NC_045512.2'):\n",
    "    Entrez.email = \"theo@theo.io\"  \n",
    "    handle = Entrez.efetch(db=\"nucleotide\", id=accession, rettype=\"fasta\", retmode=\"text\")\n",
    "    record = SeqIO.read(handle, \"fasta\")\n",
    "    handle.close()\n",
    "    return str(record.seq)\n",
    "reference_genome = fetch_reference_genome(\"NC_045512.2\")  # SARS-CoV-2 reference genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "25b0dbde",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    G>A,C>T,C>T,C>T,C>T,C>T,C>T,G>A,C>T,A>C,A>T,G>...\n",
       "1                                                     \n",
       "2                                                     \n",
       "3                                                  G>T\n",
       "4                                              G>T,A>G\n",
       "Name: subs, dtype: object"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_mut_type(mut_string):\n",
    "    if isinstance(mut_string, str) and len(mut_string) >= 2:\n",
    "        return mut_string[0] + '>' + mut_string[-1]\n",
    "    else:\n",
    "        return ''\n",
    "df[\"subs\"]=df[\"privateNucMutations.unlabeledSubstitutions\"].apply(lambda x: ','.join([get_mut_type(item) for item in x.split(',')])if isinstance(x, str) else '')\n",
    "df[\"subs\"].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cbd1343",
   "metadata": {},
   "outputs": [],
   "source": [
    "df[\"Counts\"] =df[\"subs\"].apply(lambda x: dict(Counter(x.split(\",\"))))\n",
    "df[\"Counts\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d25508fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_all_trinucleotide_contexts(genome):\n",
    "    \"\"\"Count all trinucleotide contexts in genome\"\"\"\n",
    "    counts = Counter()\n",
    "    for i in range(len(genome)-2):\n",
    "        context = genome[i:i+3]\n",
    "        counts[context] += 1\n",
    "    return counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6e40e7e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      CGT,ACT,TCT,ACC,CCT,GCT,TCA,AGG,TCA,GAG,CAA,TG...\n",
       "1                                                       \n",
       "2                                                       \n",
       "3                                                    AGT\n",
       "4                                                TGG,AAT\n",
       "                             ...                        \n",
       "994                                                  AAG\n",
       "995                                          TCT,ACA,CTA\n",
       "996                      AGT,AGT,ATA,ATG,TGT,AGT,ATA,TCA\n",
       "997                          TGG,TCC,ACT,TGC,GAA,GGA,GAA\n",
       "998                                      GCT,AAA,GGG,CGA\n",
       "Name: context, Length: 999, dtype: object"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_context(genome_seq, mutation):\n",
    "    if isinstance(mutation, str) and len(mutation) >= 2:\n",
    "        pos = int(mutation[1:-1]) - 1  # -1 as Python uses 0-based indexing\n",
    "        context = genome_seq[pos-1:pos+2]  # get the base before and after\n",
    "        return context\n",
    "    else:\n",
    "        return ''\n",
    "df[\"context\"]=df[\"privateNucMutations.unlabeledSubstitutions\"].apply(lambda x: ','.join([get_context(reference_genome, item) for item in x.split(',')])if isinstance(x, str) else '')\n",
    "df[\"context\"]   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "464cd772",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    C[G>A]T,A[C>T]T,T[C>T]T,A[C>T]C,C[C>T]T,G[C>T]...\n",
       "1                                                     \n",
       "2                                                     \n",
       "3                                              A[G>T]T\n",
       "4                                      T[G>T]G,A[A>G]T\n",
       "Name: spectrum, dtype: object"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def spectrum(subs, contexts):\n",
    "    if not isinstance(subs, str) or not isinstance(contexts, str):\n",
    "        return ''\n",
    "    if not subs.strip() or not contexts.strip():\n",
    "        return ''\n",
    "    subs= subs.split(',')\n",
    "    contexts = contexts.split(',')\n",
    "    spectra = []\n",
    "    for mutation, context in zip(subs, contexts): \n",
    "        spectra.append(f\"{context[0]}[{mutation}]{context[-1]}\")\n",
    "    return ','.join(spectra)\n",
    "   \n",
    "df[\"spectrum\"] = df.apply(lambda row: spectrum(row[\"subs\"], row[\"context\"]), axis=1)\n",
    "df[\"spectrum\"].head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a16a18b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      {'C[G>A]T': 1, 'A[G>A]G': 1, 'T[G>A]C': 1}\n",
       "1                                              {}\n",
       "2                                              {}\n",
       "3                                              {}\n",
       "4                                              {}\n",
       "                          ...                    \n",
       "994                                            {}\n",
       "995                                            {}\n",
       "996                                {'T[G>A]T': 1}\n",
       "997                  {'T[G>A]G': 1, 'G[G>A]A': 1}\n",
       "998                                {'G[G>A]G': 1}\n",
       "Name: context_counts, Length: 999, dtype: object"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def count_GtoA(spectrum):\n",
    "    counts = Counter()\n",
    "    muts = spectrum.split(\",\")\n",
    "    for mut in muts:\n",
    "        if mut[2:5] == 'G>A':\n",
    "            counts[mut]+=1\n",
    "    return counts\n",
    "\n",
    "df[\"context_counts\"] = df[\"spectrum\"].apply(count_GtoA)\n",
    "df[\"context_counts\"]\n",
    "\n",
    "            \n",
    "\n",
    "            \n",
    "\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 300,
   "id": "ff7d01dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['A[G>A]A',\n",
       " 'A[G>A]C',\n",
       " 'A[G>A]G',\n",
       " 'A[G>A]T',\n",
       " 'C[G>A]A',\n",
       " 'C[G>A]C',\n",
       " 'C[G>A]G',\n",
       " 'C[G>A]T',\n",
       " 'G[G>A]A',\n",
       " 'G[G>A]C',\n",
       " 'G[G>A]G',\n",
       " 'G[G>A]T',\n",
       " 'T[G>A]A',\n",
       " 'T[G>A]C',\n",
       " 'T[G>A]G',\n",
       " 'T[G>A]T']"
      ]
     },
     "execution_count": 300,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bases = [\"A\",\"C\",\"G\",\"T\"]\n",
    "possible_contexts = []\n",
    "for a in bases:\n",
    "    for b in bases:\n",
    "        possible_contexts.append(f\"{a}[G>A]{b}\")\n",
    "possible_contexts\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e71ec7f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    Counter({'C>T': 8, 'G>A': 3, 'A>G': 3, 'A>C': ...\n",
       "1                                            Counter()\n",
       "2                                            Counter()\n",
       "3                                  Counter({'G>T': 1})\n",
       "4                        Counter({'G>T': 1, 'A>G': 1})\n",
       "Name: Counts, dtype: object"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_proportion(df):\n",
    "    dict = {}\n",
    "    total = sum(df.values())\n",
    "    for key, value in df.items():\n",
    "        dict[key] = value/total\n",
    "       \n",
    "    return dict\n",
    "\n",
    "df[\"proportions\"] = df.apply(lambda row: get_proportion(row[\"context_counts\"]), axis=1)\n",
    "#df[\"proportions\"].head()\n",
    "df[\"Counts\"].head()\n",
    "\n",
    "    \n",
    "\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 292,
   "id": "ac15c719",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0     {'C>T': 8, 'G>A': 3, 'A>G': 3, 'A>C': 1, 'A>T'...\n",
      "1                                                    {}\n",
      "2                                                    {}\n",
      "3                                            {'G>T': 1}\n",
      "4                                  {'G>T': 1, 'A>G': 1}\n",
      "5                        {'T>C': 3, 'C>T': 2, 'A>G': 2}\n",
      "6                                  {'G>T': 2, 'G>A': 1}\n",
      "7                                  {'C>T': 4, 'A>G': 1}\n",
      "8              {'C>T': 2, 'T>C': 1, 'G>C': 1, 'C>A': 1}\n",
      "9                        {'C>T': 1, 'T>A': 1, 'T>G': 1}\n",
      "10    {'C>T': 7, 'T>C': 2, 'A>G': 1, 'A>T': 1, 'G>T'...\n",
      "11                       {'C>T': 4, 'G>T': 2, 'G>A': 1}\n",
      "12             {'C>T': 4, 'G>T': 3, 'T>G': 2, 'G>A': 1}\n",
      "13                                 {'C>T': 1, 'A>T': 1}\n",
      "14                                           {'C>T': 1}\n",
      "15                                                   {}\n",
      "16                                 {'C>T': 4, 'A>G': 1}\n",
      "17             {'C>T': 2, 'C>A': 2, 'G>T': 2, 'G>A': 1}\n",
      "18                       {'C>T': 4, 'G>A': 1, 'T>C': 1}\n",
      "19             {'C>T': 1, 'C>A': 1, 'A>C': 1, 'T>C': 1}\n",
      "Name: Counts, dtype: object\n",
      "<class 'pandas.core.series.Series'>\n"
     ]
    }
   ],
   "source": [
    "df[\"Counts\"] = (\n",
    "    df[\"Counts\"]\n",
    "    .astype(str)\n",
    "    .str.replace(\"Counter(\", \"\")  # remove \"Counter(\"\n",
    "    .str.rstrip(\")\")                           # remove trailing \")\"\n",
    "    .apply(lambda x: ast.literal_eval(x) if x not in [\"\", \"nan\", \"Counter\"] else {})\n",
    ")\n",
    "print(df[\"Counts\"].head(20))\n",
    "print(type(df[\"Counts\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 293,
   "id": "be9748a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    -0.378720\n",
       "1     0.000000\n",
       "2     0.000000\n",
       "3    -1.509730\n",
       "4    -1.720178\n",
       "5    -1.135238\n",
       "6    -2.514288\n",
       "7     1.534272\n",
       "8    -2.229715\n",
       "9    -2.023665\n",
       "10   -1.234451\n",
       "11   -0.769569\n",
       "12   -4.487829\n",
       "13   -1.073550\n",
       "14    0.436180\n",
       "15    0.000000\n",
       "16    1.534272\n",
       "17   -4.215102\n",
       "18    1.720991\n",
       "19   -3.112182\n",
       "Name: LLR, dtype: float64"
      ]
     },
     "execution_count": 293,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_likelihood_ratio(counts,pM,pN):\n",
    "    counts=np.array(counts,dtype=float)\n",
    "    llM= float(multinomial.logpmf(counts, n=np.sum(counts), p=pM))\n",
    "    llN = float(multinomial.logpmf(counts, n=np.sum(counts), p=pN))\n",
    "    llr=llM-llN\n",
    "    return llr\n",
    "probs_df = pd.read_csv(\"/Users/reem/Downloads/estimated_mutation_distribution.tsv\", delimiter=\"\\t\")\n",
    "#probs_df\n",
    "\n",
    "pM=probs_df[\"Molnupiravir\"].values.tolist()\n",
    "pN=probs_df[\"Normal\"].values.tolist()\n",
    "mutation_types = probs_df[\"MutationType\"].str.replace(\"→\", \">\").tolist()\n",
    "\n",
    "\n",
    "llr_list=[]\n",
    "for i,row in df.iterrows():\n",
    "    counts_dict = row[\"Counts\"]\n",
    "    counts = [counts_dict.get(mut,0) for mut in mutation_types]\n",
    "    llr = get_likelihood_ratio(counts,pM,pN)\n",
    "    llr_list.append(llr)\n",
    "df[\"LLR\"] = llr_list\n",
    "df[\"LLR\"].head(20)\n",
    "\n",
    "# sample_counter = df[\"Counts\"].iloc[0]\n",
    "# print(sample_counter)\n",
    "# test_row = df[\"Counts\"].iloc[0]\n",
    "# print(test_row)\n",
    "# counts_dict = test_row\n",
    "# counts = np.array([counts_dict.get(mt, 0) for mt in mutation_types])\n",
    "# print(counts)\n",
    "\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "380721af",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mNameError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[1]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m \u001b[43mdf\u001b[49m.head()\n",
      "\u001b[31mNameError\u001b[39m: name 'df' is not defined"
     ]
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c597b0d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>index</th>\n",
       "      <th>seqName</th>\n",
       "      <th>clade</th>\n",
       "      <th>clade_display</th>\n",
       "      <th>clade_who</th>\n",
       "      <th>clade_nextstrain</th>\n",
       "      <th>partiallyAliased</th>\n",
       "      <th>Nextclade_pango</th>\n",
       "      <th>qc.overallScore</th>\n",
       "      <th>...</th>\n",
       "      <th>warnings</th>\n",
       "      <th>errors</th>\n",
       "      <th>subs</th>\n",
       "      <th>Counts</th>\n",
       "      <th>LLR</th>\n",
       "      <th>Country</th>\n",
       "      <th>context</th>\n",
       "      <th>spectrum</th>\n",
       "      <th>context_counts</th>\n",
       "      <th>proportions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>9</td>\n",
       "      <td>hCoV-19/Germany/NW-RKI-I-854767/2022|2022-02-0...</td>\n",
       "      <td>21K</td>\n",
       "      <td>21K (BA.1)</td>\n",
       "      <td>Omicron</td>\n",
       "      <td>21K</td>\n",
       "      <td>BA.1.1</td>\n",
       "      <td>BA.1.1</td>\n",
       "      <td>23.390864</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C&gt;T,C&gt;T,A&gt;G,C&gt;T,C&gt;T</td>\n",
       "      <td>{'C&gt;T': 4, 'A&gt;G': 1}</td>\n",
       "      <td>1.534272</td>\n",
       "      <td>Germany</td>\n",
       "      <td>TCA,TCC,GAA,TCA,GCA</td>\n",
       "      <td>T[C&gt;T]A,T[C&gt;T]C,G[A&gt;G]A,T[C&gt;T]A,G[C&gt;T]A</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>16</td>\n",
       "      <td>17</td>\n",
       "      <td>hCoV-19/Canada/ON-KHS-22-02511-v1/2022|2022-03...</td>\n",
       "      <td>21K</td>\n",
       "      <td>21K (BA.1)</td>\n",
       "      <td>Omicron</td>\n",
       "      <td>21K</td>\n",
       "      <td>BA.1.1</td>\n",
       "      <td>BA.1.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C&gt;T,A&gt;G,C&gt;T,C&gt;T,C&gt;T</td>\n",
       "      <td>{'C&gt;T': 4, 'A&gt;G': 1}</td>\n",
       "      <td>1.534272</td>\n",
       "      <td>Canada</td>\n",
       "      <td>TCA,TAG,ACA,TCT,TCT</td>\n",
       "      <td>T[C&gt;T]A,T[A&gt;G]G,A[C&gt;T]A,T[C&gt;T]T,T[C&gt;T]T</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>18</td>\n",
       "      <td>16</td>\n",
       "      <td>hCoV-19/USA/HI-H2327938/2024|2024-01-30|2024-0...</td>\n",
       "      <td>24A</td>\n",
       "      <td>24A (JN.1)</td>\n",
       "      <td>Omicron</td>\n",
       "      <td>24A</td>\n",
       "      <td>BA.2.86.1.1.4</td>\n",
       "      <td>JN.1.4</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G&gt;A,C&gt;T,C&gt;T,C&gt;T,C&gt;T,T&gt;C</td>\n",
       "      <td>{'C&gt;T': 4, 'G&gt;A': 1, 'T&gt;C': 1}</td>\n",
       "      <td>1.720991</td>\n",
       "      <td>USA</td>\n",
       "      <td>TGG,ACA,TCC,TCT,ACA,GTT</td>\n",
       "      <td>T[G&gt;A]G,A[C&gt;T]A,T[C&gt;T]C,T[C&gt;T]T,A[C&gt;T]A,G[T&gt;C]T</td>\n",
       "      <td>{'T[G&gt;A]G': 1}</td>\n",
       "      <td>{'T[G&gt;A]G': 1.0}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>26</td>\n",
       "      <td>23</td>\n",
       "      <td>hCoV-19/England/QEUH-9E4607/2020|2020-09-23|20...</td>\n",
       "      <td>20A</td>\n",
       "      <td>20A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20A</td>\n",
       "      <td>B.1.416.1</td>\n",
       "      <td>B.1.416.1</td>\n",
       "      <td>0.173611</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C&gt;T,C&gt;T,C&gt;T,G&gt;A,C&gt;T</td>\n",
       "      <td>{'C&gt;T': 4, 'G&gt;A': 1}</td>\n",
       "      <td>2.249892</td>\n",
       "      <td>England</td>\n",
       "      <td>ACA,ACA,ACT,GGT,ACG</td>\n",
       "      <td>A[C&gt;T]A,A[C&gt;T]A,A[C&gt;T]T,G[G&gt;A]T,A[C&gt;T]G</td>\n",
       "      <td>{'G[G&gt;A]T': 1}</td>\n",
       "      <td>{'G[G&gt;A]T': 1.0}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>38</td>\n",
       "      <td>44</td>\n",
       "      <td>hCoV-19/USA/CA-CDC-FG-148485/2021|2021-10-12|2...</td>\n",
       "      <td>21J</td>\n",
       "      <td>21J (Delta)</td>\n",
       "      <td>Delta</td>\n",
       "      <td>21J</td>\n",
       "      <td>B.1.617.2.103</td>\n",
       "      <td>AY.103</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C&gt;T,C&gt;T,G&gt;A</td>\n",
       "      <td>{'C&gt;T': 2, 'G&gt;A': 1}</td>\n",
       "      <td>1.377532</td>\n",
       "      <td>USA</td>\n",
       "      <td>GCC,CCT,AGC</td>\n",
       "      <td>G[C&gt;T]C,C[C&gt;T]T,A[G&gt;A]C</td>\n",
       "      <td>{'A[G&gt;A]C': 1}</td>\n",
       "      <td>{'A[G&gt;A]C': 1.0}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 111 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Unnamed: 0  index                                            seqName  \\\n",
       "7            7      9  hCoV-19/Germany/NW-RKI-I-854767/2022|2022-02-0...   \n",
       "16          16     17  hCoV-19/Canada/ON-KHS-22-02511-v1/2022|2022-03...   \n",
       "18          18     16  hCoV-19/USA/HI-H2327938/2024|2024-01-30|2024-0...   \n",
       "26          26     23  hCoV-19/England/QEUH-9E4607/2020|2020-09-23|20...   \n",
       "38          38     44  hCoV-19/USA/CA-CDC-FG-148485/2021|2021-10-12|2...   \n",
       "\n",
       "   clade clade_display clade_who clade_nextstrain partiallyAliased  \\\n",
       "7    21K    21K (BA.1)   Omicron              21K           BA.1.1   \n",
       "16   21K    21K (BA.1)   Omicron              21K           BA.1.1   \n",
       "18   24A    24A (JN.1)   Omicron              24A    BA.2.86.1.1.4   \n",
       "26   20A           20A       NaN              20A        B.1.416.1   \n",
       "38   21J   21J (Delta)     Delta              21J    B.1.617.2.103   \n",
       "\n",
       "   Nextclade_pango  qc.overallScore  ... warnings  errors  \\\n",
       "7           BA.1.1        23.390864  ...      NaN     NaN   \n",
       "16          BA.1.1         0.000000  ...      NaN     NaN   \n",
       "18          JN.1.4         0.000000  ...      NaN     NaN   \n",
       "26       B.1.416.1         0.173611  ...      NaN     NaN   \n",
       "38          AY.103         0.000000  ...      NaN     NaN   \n",
       "\n",
       "                       subs                          Counts       LLR  \\\n",
       "7       C>T,C>T,A>G,C>T,C>T            {'C>T': 4, 'A>G': 1}  1.534272   \n",
       "16      C>T,A>G,C>T,C>T,C>T            {'C>T': 4, 'A>G': 1}  1.534272   \n",
       "18  G>A,C>T,C>T,C>T,C>T,T>C  {'C>T': 4, 'G>A': 1, 'T>C': 1}  1.720991   \n",
       "26      C>T,C>T,C>T,G>A,C>T            {'C>T': 4, 'G>A': 1}  2.249892   \n",
       "38              C>T,C>T,G>A            {'C>T': 2, 'G>A': 1}  1.377532   \n",
       "\n",
       "    Country                  context  \\\n",
       "7   Germany      TCA,TCC,GAA,TCA,GCA   \n",
       "16   Canada      TCA,TAG,ACA,TCT,TCT   \n",
       "18      USA  TGG,ACA,TCC,TCT,ACA,GTT   \n",
       "26  England      ACA,ACA,ACT,GGT,ACG   \n",
       "38      USA              GCC,CCT,AGC   \n",
       "\n",
       "                                           spectrum  context_counts  \\\n",
       "7           T[C>T]A,T[C>T]C,G[A>G]A,T[C>T]A,G[C>T]A              {}   \n",
       "16          T[C>T]A,T[A>G]G,A[C>T]A,T[C>T]T,T[C>T]T              {}   \n",
       "18  T[G>A]G,A[C>T]A,T[C>T]C,T[C>T]T,A[C>T]A,G[T>C]T  {'T[G>A]G': 1}   \n",
       "26          A[C>T]A,A[C>T]A,A[C>T]T,G[G>A]T,A[C>T]G  {'G[G>A]T': 1}   \n",
       "38                          G[C>T]C,C[C>T]T,A[G>A]C  {'A[G>A]C': 1}   \n",
       "\n",
       "         proportions  \n",
       "7                 {}  \n",
       "16                {}  \n",
       "18  {'T[G>A]G': 1.0}  \n",
       "26  {'G[G>A]T': 1.0}  \n",
       "38  {'A[G>A]C': 1.0}  \n",
       "\n",
       "[5 rows x 111 columns]"
      ]
     },
     "execution_count": 296,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_Mov = df[df[\"LLR\"]>6]\n",
    "df_Mov.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 309,
   "id": "54ba90ae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>C[G&gt;A]T</th>\n",
       "      <th>A[G&gt;A]G</th>\n",
       "      <th>T[G&gt;A]C</th>\n",
       "      <th>A[G&gt;A]C</th>\n",
       "      <th>C[G&gt;A]G</th>\n",
       "      <th>T[G&gt;A]G</th>\n",
       "      <th>G[G&gt;A]T</th>\n",
       "      <th>A[G&gt;A]A</th>\n",
       "      <th>T[G&gt;A]T</th>\n",
       "      <th>A[G&gt;A]T</th>\n",
       "      <th>G[G&gt;A]C</th>\n",
       "      <th>G[G&gt;A]G</th>\n",
       "      <th>T[G&gt;A]A</th>\n",
       "      <th>C[G&gt;A]C</th>\n",
       "      <th>G[G&gt;A]A</th>\n",
       "      <th>C[G&gt;A]A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>994</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>996</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>997</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>999 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      C[G>A]T   A[G>A]G   T[G>A]C  A[G>A]C  C[G>A]G  T[G>A]G  G[G>A]T  \\\n",
       "0    0.333333  0.333333  0.333333      NaN      NaN      NaN      NaN   \n",
       "1         NaN       NaN       NaN      NaN      NaN      NaN      NaN   \n",
       "2         NaN       NaN       NaN      NaN      NaN      NaN      NaN   \n",
       "3         NaN       NaN       NaN      NaN      NaN      NaN      NaN   \n",
       "4         NaN       NaN       NaN      NaN      NaN      NaN      NaN   \n",
       "..        ...       ...       ...      ...      ...      ...      ...   \n",
       "994       NaN       NaN       NaN      NaN      NaN      NaN      NaN   \n",
       "995       NaN       NaN       NaN      NaN      NaN      NaN      NaN   \n",
       "996       NaN       NaN       NaN      NaN      NaN      NaN      NaN   \n",
       "997       NaN       NaN       NaN      NaN      NaN      0.5      NaN   \n",
       "998       NaN       NaN       NaN      NaN      NaN      NaN      NaN   \n",
       "\n",
       "     A[G>A]A  T[G>A]T  A[G>A]T  G[G>A]C  G[G>A]G  T[G>A]A  C[G>A]C  G[G>A]A  \\\n",
       "0        NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN   \n",
       "1        NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN   \n",
       "2        NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN   \n",
       "3        NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN   \n",
       "4        NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN   \n",
       "..       ...      ...      ...      ...      ...      ...      ...      ...   \n",
       "994      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN   \n",
       "995      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN   \n",
       "996      NaN      1.0      NaN      NaN      NaN      NaN      NaN      NaN   \n",
       "997      NaN      NaN      NaN      NaN      NaN      NaN      NaN      0.5   \n",
       "998      NaN      NaN      NaN      NaN      1.0      NaN      NaN      NaN   \n",
       "\n",
       "     C[G>A]A  \n",
       "0        NaN  \n",
       "1        NaN  \n",
       "2        NaN  \n",
       "3        NaN  \n",
       "4        NaN  \n",
       "..       ...  \n",
       "994      NaN  \n",
       "995      NaN  \n",
       "996      NaN  \n",
       "997      NaN  \n",
       "998      NaN  \n",
       "\n",
       "[999 rows x 16 columns]"
      ]
     },
     "execution_count": 309,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proportions_list = pd.json_normalize(df['proportions'])\n",
    "proportions_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b898b8f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 341,
   "id": "734a483c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqName</th>\n",
       "      <th>C[G&gt;A]T</th>\n",
       "      <th>A[G&gt;A]G</th>\n",
       "      <th>T[G&gt;A]C</th>\n",
       "      <th>A[G&gt;A]C</th>\n",
       "      <th>C[G&gt;A]G</th>\n",
       "      <th>T[G&gt;A]G</th>\n",
       "      <th>G[G&gt;A]T</th>\n",
       "      <th>A[G&gt;A]A</th>\n",
       "      <th>T[G&gt;A]T</th>\n",
       "      <th>A[G&gt;A]T</th>\n",
       "      <th>G[G&gt;A]C</th>\n",
       "      <th>G[G&gt;A]G</th>\n",
       "      <th>T[G&gt;A]A</th>\n",
       "      <th>C[G&gt;A]C</th>\n",
       "      <th>G[G&gt;A]A</th>\n",
       "      <th>C[G&gt;A]A</th>\n",
       "      <th>LLR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hCoV-19/USA/MO-WRAIR-COX5040NPS/2020|2020-08-1...</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.378720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hCoV-19/Belgium/UGent-14493/2021|2021-12-22|20...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hCoV-19/France/IDF-HMN-21052200412/2021|2021-0...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hCoV-19/England/LSPA-37EF052/2022|2022-02-21|2...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.509730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hCoV-19/Germany/BE-RKI-I-595719/2022|2022-02-2...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.720178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>hCoV-19/Russia/TAM-RII-MH144950S/2023|2023-09-...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.135238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>hCoV-19/Canada/ON-KHS-21-06687-v1/2021|2021-12...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-2.514288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>hCoV-19/Germany/NW-RKI-I-854767/2022|2022-02-0...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.534272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>hCoV-19/England/ALDP-1FEE90B/2021|2021-09-29|2...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-2.229715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>hCoV-19/Northern Mariana Islands/MP-CDC-2-5839...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-2.023665</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             seqName   C[G>A]T   A[G>A]G  \\\n",
       "0  hCoV-19/USA/MO-WRAIR-COX5040NPS/2020|2020-08-1...  0.333333  0.333333   \n",
       "1  hCoV-19/Belgium/UGent-14493/2021|2021-12-22|20...  0.000000  0.000000   \n",
       "2  hCoV-19/France/IDF-HMN-21052200412/2021|2021-0...  0.000000  0.000000   \n",
       "3  hCoV-19/England/LSPA-37EF052/2022|2022-02-21|2...  0.000000  0.000000   \n",
       "4  hCoV-19/Germany/BE-RKI-I-595719/2022|2022-02-2...  0.000000  0.000000   \n",
       "5  hCoV-19/Russia/TAM-RII-MH144950S/2023|2023-09-...  0.000000  0.000000   \n",
       "6  hCoV-19/Canada/ON-KHS-21-06687-v1/2021|2021-12...  0.000000  0.000000   \n",
       "7  hCoV-19/Germany/NW-RKI-I-854767/2022|2022-02-0...  0.000000  0.000000   \n",
       "8  hCoV-19/England/ALDP-1FEE90B/2021|2021-09-29|2...  0.000000  0.000000   \n",
       "9  hCoV-19/Northern Mariana Islands/MP-CDC-2-5839...  0.000000  0.000000   \n",
       "\n",
       "    T[G>A]C  A[G>A]C  C[G>A]G  T[G>A]G  G[G>A]T  A[G>A]A  T[G>A]T  A[G>A]T  \\\n",
       "0  0.333333      0.0      0.0      0.0      0.0      0.0      0.0      0.0   \n",
       "1  0.000000      0.0      0.0      0.0      0.0      0.0      0.0      0.0   \n",
       "2  0.000000      0.0      0.0      0.0      0.0      0.0      0.0      0.0   \n",
       "3  0.000000      0.0      0.0      0.0      0.0      0.0      0.0      0.0   \n",
       "4  0.000000      0.0      0.0      0.0      0.0      0.0      0.0      0.0   \n",
       "5  0.000000      0.0      0.0      0.0      0.0      0.0      0.0      0.0   \n",
       "6  0.000000      1.0      0.0      0.0      0.0      0.0      0.0      0.0   \n",
       "7  0.000000      0.0      0.0      0.0      0.0      0.0      0.0      0.0   \n",
       "8  0.000000      0.0      0.0      0.0      0.0      0.0      0.0      0.0   \n",
       "9  0.000000      0.0      0.0      0.0      0.0      0.0      0.0      0.0   \n",
       "\n",
       "   G[G>A]C  G[G>A]G  T[G>A]A  C[G>A]C  G[G>A]A  C[G>A]A       LLR  \n",
       "0      0.0      0.0      0.0      0.0      0.0      0.0 -0.378720  \n",
       "1      0.0      0.0      0.0      0.0      0.0      0.0  0.000000  \n",
       "2      0.0      0.0      0.0      0.0      0.0      0.0  0.000000  \n",
       "3      0.0      0.0      0.0      0.0      0.0      0.0 -1.509730  \n",
       "4      0.0      0.0      0.0      0.0      0.0      0.0 -1.720178  \n",
       "5      0.0      0.0      0.0      0.0      0.0      0.0 -1.135238  \n",
       "6      0.0      0.0      0.0      0.0      0.0      0.0 -2.514288  \n",
       "7      0.0      0.0      0.0      0.0      0.0      0.0  1.534272  \n",
       "8      0.0      0.0      0.0      0.0      0.0      0.0 -2.229715  \n",
       "9      0.0      0.0      0.0      0.0      0.0      0.0 -2.023665  "
      ]
     },
     "execution_count": 341,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pivot = pd.concat([df[[\"seqName\"]],proportions_list,df[[\"LLR\"]]], axis=1)\n",
    "df_pivot.fillna(0, inplace=True)\n",
    "df_pivot.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 362,
   "id": "4bca40f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     seqName   C[G>A]T  A[G>A]G   T[G>A]C   A[G>A]C   C[G>A]G  T[G>A]G  \\\n",
      "1086     NaN       NaN      NaN       NaN       NaN       NaN      NaN   \n",
      "1087     NaN       NaN      NaN       NaN       NaN       NaN      NaN   \n",
      "1088     NaN       NaN      NaN       NaN       NaN       NaN      NaN   \n",
      "1089     NaN       NaN      NaN       NaN       NaN       NaN      NaN   \n",
      "1090     NaN       NaN      NaN       NaN       NaN       NaN      NaN   \n",
      "1091     NaN       NaN      NaN       NaN       NaN       NaN      NaN   \n",
      "1092     NaN       NaN      NaN       NaN       NaN       NaN      NaN   \n",
      "1093     NaN       NaN      NaN       NaN       NaN       NaN      NaN   \n",
      "1094     NaN       NaN      NaN       NaN       NaN       NaN      NaN   \n",
      "1095    Mean  0.020187  0.02631  0.018635  0.017327  0.009843  0.02711   \n",
      "\n",
      "       G[G>A]T   A[G>A]A   T[G>A]T   A[G>A]T  G[G>A]C   G[G>A]G   T[G>A]A  \\\n",
      "1086       NaN       NaN       NaN       NaN      NaN       NaN       NaN   \n",
      "1087       NaN       NaN       NaN       NaN      NaN       NaN       NaN   \n",
      "1088       NaN       NaN       NaN       NaN      NaN       NaN       NaN   \n",
      "1089       NaN       NaN       NaN       NaN      NaN       NaN       NaN   \n",
      "1090       NaN       NaN       NaN       NaN      NaN       NaN       NaN   \n",
      "1091       NaN       NaN       NaN       NaN      NaN       NaN       NaN   \n",
      "1092       NaN       NaN       NaN       NaN      NaN       NaN       NaN   \n",
      "1093       NaN       NaN       NaN       NaN      NaN       NaN       NaN   \n",
      "1094       NaN       NaN       NaN       NaN      NaN       NaN       NaN   \n",
      "1095  0.014357  0.023452  0.034153  0.024525  0.01368  0.007508  0.018435   \n",
      "\n",
      "       C[G>A]C   G[G>A]A   C[G>A]A   LLR         0  \n",
      "1086       NaN       NaN       NaN   NaN  0.023452  \n",
      "1087       NaN       NaN       NaN   NaN  0.034153  \n",
      "1088       NaN       NaN       NaN   NaN  0.024525  \n",
      "1089       NaN       NaN       NaN   NaN  0.013680  \n",
      "1090       NaN       NaN       NaN   NaN  0.007508  \n",
      "1091       NaN       NaN       NaN   NaN  0.018435  \n",
      "1092       NaN       NaN       NaN   NaN  0.009593  \n",
      "1093       NaN       NaN       NaN   NaN  0.014464  \n",
      "1094       NaN       NaN       NaN   NaN  0.007708  \n",
      "1095  0.009593  0.014464  0.007708  None  0.017955  \n"
     ]
    }
   ],
   "source": [
    "# context_cols = [col for col in df_pivot.columns if col not in [\"seqName\",\"LLR\"]]\n",
    "# mean_row = df_pivot[context_cols].mean()\n",
    "mean_row = df_pivot.drop(columns = [\"seqName\",\"LLR\"]).mean()\n",
    "mean_row[\"seqName\"] = \"Mean\"\n",
    "mean_row[\"LLR\"] = \"None\"\n",
    "mean_row = mean_row[df_pivot.columns]\n",
    "df = pd.concat([df_pivot,pd.DataFrame([mean_row])],ignore_index=True)\n",
    "print(df.tail(10))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68fdca90",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
